packages/server/api/src/app/mcp/mcp-server/mcp-server.ts

```diff
@@ -253,6 +269,135 @@ async function addFlowToServer(
 
 }
 
+
+async function initializeOpenAIModel({
+    platformId,
+    projectId,
+}: InitializeOpenAIModelParams): Promise<LanguageModelV1> {
+    const model = 'gpt-4o-mini'
+    const baseUrl = await domainHelper.getPublicApiUrl({
+        path: '/v1/ai-providers/proxy/openai/v1/',
+        platformId,
+    })
+
+    const apiKey = await accessTokenManager.generateEngineToken({
+        platformId,
+        projectId,
+    })
+
+    return  createOpenAI({
+        baseURL: baseUrl,
+        apiKey,
+    }).chat(model)
+}
+
+async function parseActionParametersFromInstructions({
+    actionMetadata,
+    toolPieceMetadata,
+    userInstructions,    
+    piecePackage,
+    platformId,
+    projectId,
+    logger,
+}: ParseActionParametersParams): Promise<Record<string, unknown>> {
+    const pieceConnectionExternalId = `{{connections['${toolPieceMetadata.connectionExternalId}']}}`
+    assertNotNullOrUndefined(pieceConnectionExternalId, 'Tool has no connection with the piece, please try to add a connection to the tool')
+
+    const aiModel = await initializeOpenAIModel({
+        platformId,
+        projectId,
+    })
+
+    const actionProperties = actionMetadata.props
+    const actionPropertiesEntries = Object.entries(actionProperties)
+
```

Comment:
I don't think that's correct, can you explain how this is the right order to process action properties?